tests/_signing/pkcs11uri_test.py

```diff
@@ -0,0 +1,315 @@
+# Copyright (c) 2025, IBM CORPORATION.  All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import os
+import pathlib
+import shutil
+import subprocess
+import sys
+from typing import Optional
+
+import pytest
+
+import model_signing
+from model_signing._signing.pkcs11uri import Pkcs11URI
+from model_signing._signing.pkcs11uri import escape
+
+
+MODULE_PATHS = [
+    "/usr/lib64/pkcs11/",  # Fedora, RHEL, openSUSE
+    "/usr/lib/pkcs11/",  # Fedora 32 bit, ArchLinux
+    "/usr/lib/softhsm/",  # Ubuntu, Debian, Alpine
+]
+
+
+def verify_uri(uri: Pkcs11URI, expecteduri: str) -> None:
+    encoded = uri.format()
+    assert encoded == expecteduri
+
+
+def verify_pin(uri: Pkcs11URI, expectedpin: str) -> None:
+    assert uri.has_pin()
+    pin = uri.get_pin()
+    assert pin == expectedpin
+
+
+def _test_construct(uri: Pkcs11URI, expecteduri: str) -> None:
+    verify_uri(uri, expecteduri)
+
+    expectedpin = "the pin"
+    expecteduri += "?pin-value=the%20pin"
+
+    uri.add_query_attribute("pin-value", "the%20pin")
+
+    verify_uri(uri, expecteduri)
+    verify_pin(uri, expectedpin)
+
+    uri.remove_query_attribute("pin-value")
+    uri.add_query_attribute_unencoded("pin-value", expectedpin.encode("utf-8"))
+
+    verify_uri(uri, expecteduri)
+    verify_pin(uri, expectedpin)
+
+
+def test_construct1() -> None:
+    uri = Pkcs11URI()
+
+    uri.add_path_attribute("id", "%66oo")
+    _test_construct(uri, "pkcs11:id=%66%6F%6F")
+
+
+def test_construct2() -> None:
```

Comment:
Can you push these into a class and also mark as integration so we don't require all developers to install PKCS to run tests locally?

As an example: https://github.com/sigstore/model-transparency/blob/e26c26954467b5ee05f3cfc24c2c6b363338b2fc/tests/api_test.py#L85-L87

Reply:
> Can you push these into a class and also mark as integration so we don't require all developers to install PKCS to run tests locally?

Then we have to accommodate pyproject.toml as well and have a try-except around the import of the PyKCS11 module? Is it a problem if users need to install PyKCS11 now? It should be fine for as long as it compiles?

Reply:
Or is the slow testing with SoftHSM the issue?

Reply:
I assume it's the slow SoftHSM test and so I moved that into an integration test. The other tests for the URI parser run quickly.

Reply:
> Can you push these into a class and also mark as integration so we don't require all developers to install PKCS to run tests locally?

Is the installation of PyKCS11 the problem or is the slow running test with SoftHSM a concern? I addressed the latter.

Taking out PyKCS11 becomes a bigger issue with imports and type checking. I am not sure what the best way would be to solve this. For the developer I think having to install this module shouldn't be an issue if it compiles correctly. The tests do take a bit longer now since it may require C++-compilation.